workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,187 @@
+            }
+        }
+
+        public void DeInjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (invalidMonoBehaviourTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public List<uint> GetRequiredReaderComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentReaderIdsForBehaviours[behaviourType];
+        }
+
+        public List<uint> GetRequiredWriterComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentWriterIdsForBehaviours[behaviourType];
+        }
+
+        private void Inject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var readerWriter = ReaderWriterFactory.CreateReaderWriter(componentId);
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, readerWriter);
+        }
+
+        private void DeInject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, null);
+        }
+
+        private void EnsureLoaded(Type behaviourType)
+        {
+            if (adapterCache.ContainsKey(behaviourType))
+            {
+                return;
+            }
+
+            var adapters = GetMembersWithMatchingAttributes(behaviourType);
+            var componentIdsToAdapters = new Dictionary<uint, FieldInfo>();
+            var readerIds = new List<uint>();
+            var writerIds = new List<uint>();
+            adapterCache[behaviourType] = componentIdsToAdapters;
+            componentReaderIdsForBehaviours[behaviourType] = readerIds;
+            componentWriterIdsForBehaviours[behaviourType] = writerIds;
+            foreach (var adapter in adapters)
+            {
+                // Figure out if reader or writer
+                // Get component ID
+                // Store in data structures
+                Type requiredType = adapter.FieldType;
+                var isReader = Attribute.IsDefined(requiredType, typeof(ReaderInterfaceAttribute), false);
+                var isWriter = Attribute.IsDefined(requiredType, typeof(WriterInterfaceAttribute), false);
+                if (!isReader && !isWriter)
```

Comment:
Is `isReader && isWriter` possible?

Reply:
No (as in: nothing should be codegenned like that)